ee/control/actionqueue/actionqueue.go

```diff
@@ -0,0 +1,241 @@
+	actionCleanupInterval time.Duration
+	cancel                context.CancelFunc
+}
+
+type actionqueueOption func(*actionqueue)
+
+func WithLogger(logger log.Logger) actionqueueOption {
+	return func(aq *actionqueue) {
+		aq.logger = logger
+	}
+}
+
+func WithStore(store types.KVStore) actionqueueOption {
+	return func(aq *actionqueue) {
+		aq.store = store
+	}
+}
+
+func WithCleanupInterval(cleanupInterval time.Duration) actionqueueOption {
+	return func(aq *actionqueue) {
+		aq.actionCleanupInterval = cleanupInterval
+	}
+}
+
+func WithContext(ctx context.Context) actionqueueOption {
+	return func(aq *actionqueue) {
+		aq.ctx = ctx
+	}
+}
+
+func New(opts ...actionqueueOption) *actionqueue {
+	aq := &actionqueue{
+		ctx:                   context.Background(),
+		updaters:              make(map[string]updater, 0),
+		actionCleanupInterval: defaultCleanupInterval,
+		logger:                log.NewNopLogger(),
+	}
+
+	for _, opt := range opts {
+		opt(aq)
+	}
+
+	if aq.store == nil {
+		aq.store = inmemory.NewStore(aq.logger)
+	}
+
+	return aq
+}
+
+func (aq *actionqueue) Update(data io.Reader) error {
+	// We want to unmarshal each action separately, so that we don't fail to send all actions
+	// if only some are malformed.
+	var rawActionsToProcess []json.RawMessage
+	if err := json.NewDecoder(data).Decode(&rawActionsToProcess); err != nil {
+		return fmt.Errorf("failed to decode actions data: %w", err)
+	}
+
+	for _, rawAction := range rawActionsToProcess {
+		var action action
+		if err := json.Unmarshal(rawAction, &action); err != nil {
+			level.Debug(aq.logger).Log("msg", "received action in unexpected format from K2, discarding", "err", err)
+			continue
+		}
+
+		if !aq.isActionValid(action) || !aq.isActionNew(action.ID) {
+			continue
+		}
+
+		updater, err := aq.updaterForAction(action)
+		if err != nil {
+			level.Info(aq.logger).Log("msg", "getting updater for action", "error", err)
+			continue
+		}
+
+		if err := updater.Update(bytes.NewReader(rawAction)); err != nil {
+			level.Info(aq.logger).Log("msg", "failed to update with action, not marking action complete", "err", err)
+			continue
+		}
+
+		// only mark processed when updater was successful
+		action.ProcessedAt = time.Now().UTC()
+		aq.storeActionRecord(action)
+	}
+
+	return nil
+}
+
+func (aq *actionqueue) RegisterUpdater(updaterType string, updater updater) {
+	aq.updaters[updaterType] = updater
+}
+
+func (aq *actionqueue) StartCleanup() error {
+	aq.runCleanup()
+	return nil
+}
+
+func (aq *actionqueue) runCleanup() {
+	ctx, cancel := context.WithCancel(aq.ctx)
+	aq.cancel = cancel
+
+	t := time.NewTicker(aq.actionCleanupInterval)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			level.Debug(aq.logger).Log("msg", "action cleanup stopped due to context cancel")
+			return
+		case <-t.C:
+			aq.cleanupActions()
+		}
+	}
+}
+
+func (aq *actionqueue) StopCleanup(err error) {
+	aq.cancel()
+}
+
+func (aq *actionqueue) storeActionRecord(action action) {
```

Comment:
Can we rename the variable (to `actionToStore` or `a`) so that it's not the same as the type? Reusing type or package identifiers as variable names re-defines that keyword for the current scope, which can result in [confusing behavior](https://play.golang.com/p/dCoezMdRnwY).

Reply:
done